crates/nu-engine/src/script.rs

```diff
@@ -33,36 +38,119 @@ fn chomp_newline(s: &str) -> &str {
     }
 }
 
-pub fn run_script_in_dir(
-    script: String,
-    dir: &Path,
-    ctx: &EvaluationContext,
-) -> Result<(), Box<dyn Error>> {
-    //Save path before to switch back to it after executing script
-    let path_before = ctx.shell_manager.path();
+/// Runs script `script` configurable by `options`
+/// All errors are printed out.
+pub fn run_script(script: NuScript, options: &RunScriptOptions, ctx: &EvaluationContext) {
+    let code = match script.get_code() {
+        Ok(code) => code,
+        Err(e) => {
+            ctx.host.lock().print_err(e, &Text::from("".to_string()));
+            return;
+        }
+    };
+
+    if let Err(e) = setup_shell(options, ctx) {
+        ctx.host.lock().print_err(e, &Text::from("".to_string()));
+        return;
+    }
+
+    //Switch to cwd if given
+    if let Some(path) = &options.with_cwd {
+        ctx.shell_manager
+            .set_path(path.to_string_lossy().to_string());
+    }
+
+    if !options.source_script {
+        ctx.scope.enter_scope()
+    }
+
+    let line_result = process_script(&code, options, ctx);
+    evaluate_line_result(line_result, options, ctx);
+
+    if !options.source_script {
+        ctx.scope.exit_scope();
+
+        //Leave shell (undoing with_cwd)
+        ctx.shell_manager.remove_at_current();
+    } else {
+        // We are in source mode
+
+        // Save script_shell path
+        let cur_path = ctx.shell_manager.path();
+        // Leave shell (undoing with_cwd)
+        ctx.shell_manager.remove_at_current();
+        // We don't update shell which called run_script with `with_cwd` path.
+        // But if a script got sourced, and it executed cd, this path change has to be propagated
+        // to the caller shell.
+        if Some(&cur_path) != options.cwd_as_string().as_ref() {
+            ctx.shell_manager.set_path(cur_path);
+        }
+    }
+}
 
-    ctx.shell_manager
-        .set_path(dir.to_string_lossy().to_string());
-    run_script_standalone(script, false, ctx, false)?;
-    ctx.shell_manager.set_path(path_before);
+fn setup_shell(options: &RunScriptOptions, ctx: &EvaluationContext) -> Result<(), ShellError> {
+    //Switch to correct shell
+    if options.cli_mode {
+        ctx.shell_manager.enter_cli_mode()?;
+    } else {
+        ctx.shell_manager.enter_script_mode()?;
+    }
 
     Ok(())
 }
 
+fn evaluate_line_result(
```

Comment:
This is `run_script_standalone`, but renamed to what it actually does and slightly changed to use `RunScriptOptions`